solr/core/src/java/org/apache/solr/search/KeepAliveRegenerator.java

```diff
@@ -0,0 +1,593 @@
+                      } else {
+                        return id - docBase;
+                      }
+                    }
+
+                    @Override
+                    public int advance(int target) {
+                      while (nextDoc() < target) {
+                        // advance
+                      }
+                      return id == NO_MORE_DOCS ? NO_MORE_DOCS : id - docBase;
+                    }
+
+                    @Override
+                    public long cost() {
+                      return 0;
+                    }
+                  };
+            } else if (stale instanceof BitDocSet) {
+              final FixedBitSet docs = ((BitDocSet) stale).getBits();
+              disi =
+                  new DocIdSetIterator() {
+                    final int limit = segment.maxDoc + docBase;
+                    int id = docBase - 1;
+
+                    @Override
+                    public int docID() {
+                      return id == NO_MORE_DOCS ? NO_MORE_DOCS : id - docBase;
+                    }
+
+                    @Override
+                    public int nextDoc() {
+                      if (++id >= limit || (id = docs.nextSetBit(id)) >= limit) {
+                        return id = NO_MORE_DOCS;
+                      } else {
+                        return id - docBase;
+                      }
+                    }
+
+                    @Override
+                    public int advance(int target) {
+                      if (target >= limit || (id = docs.nextSetBit(target)) >= limit) {
+                        return id = NO_MORE_DOCS;
+                      } else {
+                        return id - docBase;
+                      }
+                    }
+
+                    @Override
+                    public long cost() {
+                      return 0;
+                    }
+                  };
+            } else {
+              throw new IllegalStateException();
+            }
+            return new ConstantScoreScorer(this, 1f, scoreMode, disi);
+          }
+        }
+
+        @Override
+        public boolean isCacheable(LeafReaderContext ctx) {
+          return true;
+        }
+      };
+    }
+
+    @Override
+    @SuppressWarnings("ReferenceEquality")
+    public Query rewrite(IndexReader reader) throws IOException {
+      Query rewrittenBacking = backing.rewrite(reader);
+      return rewrittenBacking == backing
+          ? this
+          : new FrankensteinQuery(rewrittenBacking, segs, stale);
+    }
+
+    @Override
+    public String toString(String field) {
+      return FrankensteinQuery.class.getSimpleName() + "{" + backing + "}";
+    }
+
+    @Override
+    public void visit(QueryVisitor visitor) {
+      backing.visit(visitor);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+      return obj == this
+          || (obj instanceof FrankensteinQuery
+              && ((FrankensteinQuery) obj).backing.equals(backing));
```

Comment:
Do we not care about `segs` for equality? I almost think that 2 `FrankensteinQuery` of the same `backing` Query should not be considered equal if the segment map is not the same?